### PR TITLE
chore(deps): upgrade react-scripts and webpack to v5

### PR DESCRIPTION
**Pull request by chenba**
Because:
 - we want to upgrade our React/CRA packages to react-scripts and webpack v5

This commit:
 - upgrades react-scripts and webpack to v5 for Settings, Payments, and Admin Panel
   - upgrades other dependencies in order to make the react-scripts and webpack upgrades work
 - hybridizes fxa-auth-client and fxa-shared into dual module format packages

The most consequential change in this patch is Webpack 5's support of "exports" in package.json.  Majority of the changes are a result of that.  With the support of "exports", it means that, without any other changes, Webpack 4 and 5 could be transpiling and bundling different files. E.g. Webpack 4 transpiling the original source while Webpack 5 loads the file in "exports" that's been transpiled by tsc.

### CODE REVIEW

File: packages/functional-tests/playwright.config.ts [hunk 1]
```diff
@@ -10,9 +10,10 @@ const CI = !!process.env.CI;
 // see .vscode/launch.json
 const DEBUG = !!process.env.DEBUG;
 const SLOWMO = parseInt(process.env.PLAYWRIGHT_SLOWMO || '0');
+const NUM_WORKERS = parseInt(process.env.PLAYWRIGHT_WORKERS || '2');
```
Comment (chenba) [inline packages/functional-tests/playwright.config.ts:13]: My dev machine's got 10 cores.

File: packages/fxa-auth-client/package.json [hunk 2]
```diff
@@ -2,20 +2,36 @@
   "name": "fxa-auth-client",
   "version": "0.0.0",
   "description": "",
-  "main": "dist/server/packages/fxa-auth-client/server.js",
+  "main": "dist/server/cjs/packages/fxa-auth-client/server.js",
+  "module": "dist/server/esm/packages/fxa-auth-client/server.js",
   "exports": {
-    ".": "./dist/server/packages/fxa-auth-client/server.js",
+    ".": {
```
Comment (chenba) [inline packages/fxa-auth-client/package.json:8]: Previously it didn't matter that fxa-auth-client was transpiled into cjs since Webpack 4 was not using "exports" and loading the original source directly for Babel to transpile.  With Webpack 5's support for "exports", it became a problem.  Settings' `import` of something from this package did not work.  A dual format package allows Settings' `import` to work while maintaining compatibility with whatever required this package to be cjs in the first place.

File: packages/fxa-payments-server/package.json [hunk 3]
```diff
@@ -54,7 +54,7 @@
     "@babel/core": "^7.15.0",
     "@fluent/bundle": "^0.18.0",
     "@fluent/langneg": "^0.6.2",
-    "@fluent/react": "^0.13.1",
+    "@fluent/react": "^0.14.1",
```
Comment (chenba) [inline packages/fxa-payments-server/package.json:57]: fluent/react incorrectly declared the package to be commonjs even though the code was in esm.  (But I didn't upgrade to latest due to https://mozilla-hub.atlassian.net/browse/FXA-7458.)

File: packages/fxa-react/package.json [hunk 4]
```diff
@@ -3,13 +3,13 @@
...
   "exports": {
-    "./components/": "./dist/packages/fxa-react/components/",
+    "./components/": "./components/",
     "./configs/tailwind": "./configs/tailwind.js",
     "./configs/rescripts": "./configs/rescripts.js",
     "./configs/storybooks": "./configs/storybooks.js",
     "./images/": "./images/",
     "./extract-imported-components": "./extract-imported-components.js",
-    "./lib/": "./dist/packages/fxa-react/lib/"
+    "./lib/": "./lib/"
```
Comment (chenba) [inline packages/fxa-react/package.json:12]: These changes effectively get Webpack 5 to do what Webpack 4 was doing: loading and transpiling the original source.

File: packages/fxa-settings/.rescriptsrc.js [hunk 5]
```diff
@@ -2,39 +2,6 @@
...
-            path: path.resolve(fs.realpathSync(__dirname), 'build'),
-          },
-        };
-      }
-
-      return newConfig;
-    },
-  },
-  permitAdditionalJSImports,
-];
```
Comment (chenba) [inline packages/fxa-settings/.rescriptsrc.js:40]: Deleted the ftl concatenation stuff because it's done through Grunt now.

File: packages/fxa-shared/package.json [hunk 6]
```diff
@@ -2,63 +2,212 @@
...
-    "./payments/iap/google-play/types": "./dist/packages/fxa-shared/payments/iap/google-play/types/index.js",
-    "./payments/stripe": "./dist/packages/fxa-shared/payments/stripe.js",
-    "./payments/stripe-firestore": "./dist/packages/fxa-shared/payments/stripe-firestore.js",
-    "./scripts/*": "./dist/packages/fxa-shared/scripts/*.js",
-    "./sentry": "./dist/packages/fxa-shared/sentry/index.js",
-    "./sentry/*": "./dist/packages/fxa-shared/sentry/*.js",
-    "./speed-trap/*": "./dist/packages/fxa-shared/speed-trap/*.js",
-    "./subscriptions/*": "./dist/packages/fxa-shared/subscriptions/*.js",
-    "./tracing/*": "./dist/packages/fxa-shared/tracing/*.js"
+    ".": {
```
Comment (chenba) [inline packages/fxa-shared/package.json:8]: Same idea as https://github.com/mozilla/fxa/pull/15474/files#r1240456297

File: packages/fxa-shared/package.json [hunk 7]
```diff
@@ -2,63 +2,212 @@
...
+    },
+    "./tracing/*": {
+      "import": "./dist/esm/packages/fxa-shared/tracing/*.js",
+      "require": "./dist/cjs/packages/fxa-shared/tracing/*.js"
+    }
   },
   "scripts": {
     "postinstall": "yarn build || true",
-    "build": "tsc --build && yarn copy-assets && yarn copy-sql",
+    "build": "tsc --build && tsc --build tsconfig.cjs.json && echo '{ \"type\": \"module\" }' > ./dist/esm/packages/fxa-shared/db/package.json && yarn copy-assets && yarn copy-sql",
```
Comment (chenba) [inline packages/fxa-shared/package.json:207]: The extra `package.json` bit is _specifically_ for the db patcher script.  Because the transpiled files have .js extensions not .mjs node think they are cjs; the package.json fixes that.  I scoped it into the /db dir since the db patcher script was the only thing with that issue (that I'm aware of).

File: packages/fxa-shared/tsconfig.base.json [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  "extends": "../../tsconfig.base.json",
+  "compilerOptions": {
+    "allowJs": true,
+    "declaration": true,
+    "emitDecoratorMetadata": true,
+    "experimentalDecorators": true,
+    "composite": true,
+    "types": ["mocha", "mozlog", "jest", "node"]
+  },
+  "include": ["./**/*", "./**/**/*", "./*", "./**/*.json"],
```
Comment (LZoog) [inline packages/fxa-shared/tsconfig.base.json:11]: Does that first include `["./**/*"]` not include the rest?

File: packages/fxa-settings/package.json [hunk 9]
```diff
@@ -149,6 +149,6 @@
     "storybook-addon-rtl": "^0.5.0",
     "style-loader": "^1.3.0",
     "tailwindcss": "^3.3.1",
-    "webpack": "^4.43.0"
+    "webpack": "^5.84.1"
```
Comment (LZoog) [inline packages/fxa-settings/package.json:152]: Interesting that `fxa-settings` already had a webpack dep when `fxa-payments-server` does not. 🤔

**Review by LZoog — COMMENTED** [review]
Looking good on first pass and things run locally🎉, not sure I have any real alternative suggestions without diving into the weeds. I guess it ended up being unnecessary to switch from Rescripts to CRACO or react-app-rewired, I'm curious how that might affect any [Nx migration](https://nx.dev/recipes/adopting-nx/migration-cra); I'll defer to @julianpoy since he's working on something related.

File: packages/fxa-auth-client/package.json (hunk 2 above)
Comment (chenba) [reply packages/fxa-auth-client/package.json:8]: > I was going to suggest possibly filing a follow up for investigating this or `fxa-shared`'s need for this sometime down the road but that'd probably be another whack-a-mole scenario with unknown amount of refactoring

By "investigating this" do you mean not needing to export both formats?

> Plus would be nice not to need `fxa-auth-client` in the future. 

What's going to replace it in gql-api?

> Not sure if `fxa-auth-client`'s (non-existent??) README / `fxa-shared`'s README is the best place for a note but could be useful as a note or comment somewhere. 

Couldn't hurt I guess.  But I suspect people open a package's package.json more often than the README.

File: packages/fxa-settings/package.json (hunk 9 above)
Comment (chenba) [reply packages/fxa-settings/package.json:163]: That was my fault with accidentally editing the wrong set of deps (dependencies vs devDependencies).  It's been fixed in a newer commit.

File: packages/fxa-shared/tsconfig.base.json (hunk 8 above)
Comment (chenba) [reply packages/fxa-shared/tsconfig.base.json:11]: hehe it came from https://github.com/mozilla/fxa/pull/15474/files#diff-0a0a6cc1daeea9323c25d6453a557da86b013b1b50019783dd9423848510c8e4L13 so I didn't look close at it.  I think you are right but I'm not sure because a) I don't know what tool or lib the patterns are passed to, and b) maybe there's a reason someone included the additional patterns?

**Comment by chenba** [comment]
Marking this as ready so reviewers can start reviewing.  CI is failing due to `The <Localized /> component was not properly wrapped in a <LocalizationProvider />.` errors; fixing those shouldn't really affect the rest of the changes.

File: packages/fxa-payments-server/src/components/PaymentErrorView/index.stories.tsx [hunk 10]
```diff
@@ -11,20 +12,25 @@ export default {
...
-              plan={SELECTED_PLAN}
-            />
-          }
-        />
-      </Routes>
-    </BrowserRouter>
+    <AppLocalizationProvider
+      baseDir="./locales"
+      userLocales={navigator.languages}
+    >
```
Comment (chenba) [inline packages/fxa-payments-server/src/components/PaymentErrorView/index.stories.tsx:18]: I'm using AppLocalizationProvider directly instead of withLocalization in some places because withLocalization causes a not very obvious type error with tsc when used in Payments and fxa-react.  So even though both `storybook` and `build-storybook` work fine (neither uses tsc), `build` fails with type errors.  I don't particularly want to go down that rabbit hole for this PR.

File: packages/fxa-auth-client/package.json (hunk 2 above)
Comment (LZoog) [reply packages/fxa-auth-client/package.json:8]: > By "investigating this" do you mean not needing to export both formats?

Yeah, would be nice one day, heh.

> > Plus would be nice not to need fxa-auth-client in the future.
>
> What's going to replace it in gql-api?

I saved this note from Ben a while back (have it in my WIP Settings architecture doc so it's not lost):

Until we have an abstracted set of business logic outside the auth-server routing code, we need to call into auth-server for most of the functions (which auth-client does). But essentially, removing the use of the auth-client in graphql-api server needs the following steps done for each use of it:
* Analyze the URL the auth-client calls to determine all db calls and logic that the route handles
* Ensure the db logic is done in db code, refactor into fxa-shared as needed
* Ensure all needed business logic for the route is handled in fxa-shared, ideally via a new FirefoxAccount class
* Switch auth-client call to appropriate business logic calls into fxa-shared

> Couldn't hurt I guess. But I suspect people open a package's package.json more often than the README.

Yeah maybe that's true, I'll leave it up to you🤷‍♀️

**Comment by julianpoy** [comment]
Any idea why I might see this error while checking out this branch?

image

**Comment by chenba** [comment]
Sorry looks like the fluent issue where anything is used outside of a `<LocalizationProvider>` is breaking the Payments UI.

**Comment by chenba** [comment]
> Sorry looks like the fluent issue where anything is used outside of a `<LocalizationProvider>` is breaking the Payments UI.

Ok, fixed.  Or at least the checkout page is rendering for me locally.

**Comment by chenba** [comment]
@julianpoy Payments should render now.

File: packages/fxa-payments-server/package.json (hunk 3 above)
Comment (chenba) [reply packages/fxa-payments-server/package.json:60]: I actually did end up updating Storybook stories and frontend tests for the Fluent upgrade.  So I suspect that we can upgrade to the latest Fluent with those changes, but I'm not going to do that in this PR.

**Review by julianpoy — APPROVED** [review]

File: packages/fxa-auth-server/package.json [hunk 11]
```diff
@@ -205,16 +212,19 @@
     "pm2": "^5.3.0",
     "prettier": "^2.3.1",
     "proxyquire": "^2.1.3",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
```
[the remark this answers is not in the document]
Comment (chenba) [reply packages/fxa-auth-server/package.json:216]: Storybook's migration tool added them.  I guess Storybook uses React to render its pages?  I was a little surprised when I found myself dealing with Storybook in auth-server, but then yeah, email templates.  You know way more about Storybook than me...does it not use React?

File: packages/fxa-settings/package.json (hunk 9 above)
Comment (chenba) [reply packages/fxa-settings/package.json:163]: eh not a silly question at all.  I don't quite understand why it was originally added either.

I think one effect this particular change has is that this newer version will be used by react-scripts since react-scripts 5's package.json specifies its webpack dep with a `^`.  iirc at least one of the deps requires a very new version of webpack 5.  Of course we could use `resolutions` in the root package.json to achieve the same thing if that's the _only_ reason for this being here.

In other words, I don't know for certain until I try deleting that line which I haven't done so at the time of writing.  😀

Comment (chenba) [reply packages/fxa-settings/package.json:163]: Yeah I don't think this explicit dep is necessary.  I'll file a follow-up to see which of the Webpack deps we can remove.  Thanks for bring it up.
